apis/console/src/lib.rs

```diff
@@ -66,6 +74,54 @@ impl<S: Syscalls, C: platform::allow_ro::Config + platform::subscribe::Config> C
         })
     }
 
+    /// Reads bytes
+    /// Reads from the device and writes to `buf`, starting from index 0.
+    /// No special guarantees about when the read stops.
+    /// Returns count of bytes written to `buf`.
+    pub fn read(buf: &mut [u8]) -> (usize, Result<(), ErrorCode>) {
+        let called = core::cell::Cell::new(Option::<(u32, u32)>::None);
+        let mut bytes_received = 0;
+        let r = share::scope::<
+            (
+                AllowRw<_, DRIVER_NUM, { allow_rw::READ }>,
+                Subscribe<_, DRIVER_NUM, { subscribe::READ }>,
+            ),
+            _,
+            _,
+        >(|handle| {
+            let (allow_rw, subscribe) = handle.split();
+            let len = buf.len();
+            S::allow_rw::<C, DRIVER_NUM, { allow_rw::READ }>(allow_rw, buf)?;
+            S::subscribe::<_, _, C, DRIVER_NUM, { subscribe::READ }>(subscribe, &called)?;
+
+            S::command(DRIVER_NUM, command::READ, len as u32, 0)
+                .to_result()
+                .map_err(|e| match called.get() {
```

Comment:
`called.get()` will always return `None` here, because upcalls are only invoked when a Yield system call is executed.